mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala

```diff
@@ -0,0 +1,574 @@
+  }
+
+  /**
+   * Builds a cluster tree from a Map of clusters
+   *
+   * @param treeMap divided clusters as a Map class
+   * @param rootIndex index you want to start
+   * @param numClusters the number of clusters you want
+   * @return a built cluster tree
+   */
+  private[clustering]
+  def buildTree(treeMap: Map[Long, ClusterTree],
+    rootIndex: Long,
+    numClusters: Int): Option[ClusterTree] = {
+
+    // if there is no index in the Map
+    if (!treeMap.contains(rootIndex)) return None
+
+    // build a cluster tree if the queue is empty or until the number of leaves clusters is enough
+    var numLeavesClusters = 1
+    val root = treeMap(rootIndex)
+    var leavesQueue = Map(rootIndex -> root)
+    while (leavesQueue.size > 0 && numLeavesClusters < numClusters) {
+      // pick up the cluster whose variance is the maximum in the queue
+      val mostScattered = leavesQueue.maxBy(_._2.variancesNorm)
+      val mostScatteredKey = mostScattered._1
+      val mostScatteredCluster = mostScattered._2
+
+      // relate the most scattered cluster to its children clusters
+      val childrenIndexes = Array(2 * mostScatteredKey, 2 * mostScatteredKey + 1)
+      if (childrenIndexes.forall(i => treeMap.contains(i))) {
+        // insert children to the most scattered cluster
+        val children = childrenIndexes.map(i => treeMap(i))
+        mostScatteredCluster.insert(children)
+
+        // calculate the local dendrogram height
+        // TODO Supports distance metrics other Euclidean distance metric
+        val metric = (bv1: BV[Double], bv2: BV[Double]) => breezeNorm(bv1 - bv2, 2.0)
+        val localHeight = children
+            .map(child => metric(child.center.toBreeze, mostScatteredCluster.center.toBreeze)).max
+        mostScatteredCluster.setLocalHeight(localHeight)
+
+        // update the queue
+        leavesQueue = leavesQueue ++ childrenIndexes.map(i => (i -> treeMap(i))).toMap
+        numLeavesClusters += 1
+      }
+
+      // remove the cluster which is involved to the cluster tree
+      leavesQueue = leavesQueue.filterNot(_ == mostScattered)
+
+      log.info(s"Total Leaves Clusters: ${numLeavesClusters} / ${numClusters}. " +
+          s"Cluster ${childrenIndexes.mkString(",")} are merged.")
+    }
+    Some(root)
+  }
+
+  /**
+   * Divides the input data
+   *
+   * @param data the pairs of cluster index and point which you want to divide
+   * @param clusters the clusters you want to divide AS a Map class
+   * @return divided clusters as Map
+   */
+  private[clustering]
+  def divide(data: RDD[(Long, BV[Double])],
+    clusters: Map[Long, ClusterTree]): Map[Long, (BV[Double], Double, BV[Double])] = {
+
+    val sc = data.sparkContext
+    val centers = clusters.map { case (idx, cluster) => (idx, cluster.center.toBreeze)}
+    var newCenters = initChildrenCenter(centers)
+    if (newCenters.size == 0) {
+      return Map.empty[Long, (BV[Double], Double, BV[Double])]
+    }
+    sc.broadcast(newCenters)
```

Comment:
This use of `sc.broadcast` has no effect because the output isn't assigned or used. Instead, you want something like val `bcNewCenters = sc.broadcast(newCenters)` and then access within the map as bcCenters.value. This appears in a few other places as well.